### PR TITLE
Add Tweedie Deviance Score Metric.

### LINKED ISSUE #73: [New metric] Deviance scores
## 🚀 Feature

Add deviance scores:
* Poisson deviance https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_poisson_deviance.html#sklearn.metrics.mean_poisson_deviance
* Gamma deviance https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_gamma_deviance.html#sklearn.metrics.mean_gamma_deviance
* Tweedie deviance https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_tweedie_deviance.html#sklearn.metrics.mean_tweedie_deviance

Use in regression tasks and for goodness-of-fit testing.

### PR DESCRIPTION
**Pull request by karthikrangasai**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

Fixes #73 .

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/references/functional.rst [hunk 1]
```diff
@@ -245,6 +245,13 @@ cosine_similarity [func]
     :noindex:
 
 
+tweedie_deviance_score [func]
```
Comment (SkafteNicki) [inline docs/source/references/functional.rst:248]: please insert in alphabetical order

File: docs/source/references/modules.rst [hunk 2]
```diff
@@ -364,6 +364,13 @@ CosineSimilarity
     :noindex:
 
 
+TweedieDevianceScore
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:367]: please insert in alphabetical order

File: torchmetrics/functional/regression/tweedie_deviance.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            raise ValueError(f"For power={power}, both 'preds' and 'targets' have to be strictly positive.")
+
+        deviance_score = 2 * (torch.log(preds / targets) + (targets / preds) - 1)
+    else:
+        term_1 = torch.pow(torch.max(targets, torch.zeros(targets.shape)), 2 - power) / ((1 - power) * (2 - power))
+        term_2 = targets * torch.pow(preds, 1 - power) / (1 - power)
+        term_3 = torch.pow(preds, 2 - power) / (2 - power)
+        deviance_score = 2 * (term_1 - term_2 + term_3)
+
+    return torch.mean(deviance_score)
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/tweedie_deviance.py:80]: most of this logic should be moved to the update function, such that `update` just returns two tensors: `sum_deviance_score` and `num_observations` and compute basically just do: `sum_deviance_score/num_observations`. This have no effect for the functional implementation, but for accumulation in the modular implementation it will save a lot of memory.

File: torchmetrics/regression/tweedie_deviance.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+        if power < 1 and power > 0:
+            raise ValueError(f"Deviance Score is not defined for power={power}.")
+
+        self.power: int = power
+
+        self.add_state("preds", [], dist_reduce_fx="cat")
+        self.add_state("targets", [], dist_reduce_fx="cat")
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:98]: this is more memory efficient
```suggestion
        self.add_state("sum_deviance_score", torch.zeros(1), dist_reduce_fx="sum")
        self.add_state("num_obs", torch.zeros(1), dist_reduce_fx="sum")
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        >>> from torchmetrics import TweedieDevianceScore
+        >>> targets = torch.tensor([1.0, 2.0, 3.0, 4.0])
+        >>> preds = torch.tensor([4.0, 3.0, 2.0, 1.0])
+        >>> deviance_score = TweedieDevianceScore(power=0)
+        >>> deviance_score(preds, targets)
+        tensor(5.)
+
+    """
+    preds: List[Tensor]
+    targets: List[Tensor]
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:76]:
```suggestion

```

File: torchmetrics/regression/tweedie_deviance.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        """Update metric states with predictions and targets.
+
+        Args:
+            preds: Predicted tensor with shape ``(N,d)``
+            targets: Ground truth tensor with shape ``(N,d)``
+        """
+        preds, targets = _tweedie_deviance_score_update(preds, targets)
+
+        self.preds.append(preds)
+        self.targets.append(targets)
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:110]:
```suggestion
        sum_deviance_score, num_obs = _tweedie_deviance_score_update(preds, targets)

        self.sum_deviance_score += sum_deciance_score
        self.num_obs += num_obs
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+    def compute(self) -> Tensor:
+        preds = dim_zero_cat(self.preds)
+        targets = dim_zero_cat(self.targets)
+        return _tweedie_deviance_score_compute(preds, targets, self.power)
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:115]:
```suggestion
        return _tweedie_deviance_score_compute(self.sum_deviance_score, self.num_obs, self.power)
```

File: torchmetrics/functional/regression/tweedie_deviance.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    Args:
+        preds: Predicted tensor
+        targets: Ground truth tensor
+    """
+    _check_same_shape(preds, targets)
+    return preds, targets
+
+
+def _tweedie_deviance_score_compute(preds: Tensor, targets: Tensor, power: int = 0) -> Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/tweedie_deviance.py:37]: I assume you copied this implementation (https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/_loss/glm_distribution.py#L248-L323) meaning that `power` can be a float

```suggestion
def _tweedie_deviance_score_compute(preds: Tensor, targets: Tensor, power: float = 0) -> Tensor:
```

File: torchmetrics/functional/regression/tweedie_deviance.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    """Computes Cosine Similarity.
+
+    Args:
+        preds: Predicted tensor
+        targets: Ground truth tensor
+        power:
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/tweedie_deviance.py:43]: please update docstring with the additional scenarios where power is not 0,1,2
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_tweedie_deviance.html#sklearn.metrics.mean_tweedie_deviance

File: torchmetrics/regression/tweedie_deviance.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+    where :math:`y` is a tensor of targets values, and :math:`\hat{y}` is a tensor of predictions.
+
+    Forward accepts
+
+    - ``preds`` (float tensor): ``(N,)``
+    - ``targets`` (float tensor): ``(N,)``
+
+    Args:
+        power:
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:50]: please update docstring with the additional scenarios where power is not 0,1,2
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_tweedie_deviance.html#sklearn.metrics.mean_tweedie_deviance

File: torchmetrics/regression/tweedie_deviance.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+    ) -> None:
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+        if power < 1 and power > 0:
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:92]:
```suggestion
        if 0 < power < 1:
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        >>> deviance_score(preds, targets)
+        tensor(5.)
+
+    """
+    preds: List[Tensor]
+    targets: List[Tensor]
+
+    def __init__(
+        self,
+        power: int = 0,
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:80]:
```suggestion
        power: float = 0,
```

**Review by SkafteNicki — COMMENTED** [review]
Hi @karthikrangasai, thanks for the PR
It is looking fairly done, but I think it can be improved (memory wise) by not storing all preds and targets in a buffer but instead some intermediate states. Please see comments, and feel free to reach out to me if it is unclear.
Also remember to add note in changelog :]

File: torchmetrics/functional/regression/tweedie_deviance.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            if torch.any(preds <= 0) or torch.any(targets <= 0):
+                raise ValueError(f"For power={power}, both 'preds' and 'targets' have to be strictly positive.")
+
+        term_1 = torch.pow(torch.max(targets, zero_tensor), 2 - power) / ((1 - power) * (2 - power))
+        term_2 = targets * torch.pow(preds, 1 - power) / (1 - power)
+        term_3 = torch.pow(preds, 2 - power) / (2 - power)
+        deviance_score = 2 * (term_1 - term_2 + term_3)
+
+    sum_deviance_score = torch.sum(deviance_score)
+    num_observations = torch.tensor(torch.numel(deviance_score))
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/tweedie_deviance.py:86]: device mismatch possible
```suggestion
    num_observations = torch.tensor(torch.numel(deviance_score), device=preds.device)
```

File: torchmetrics/functional/regression/tweedie_deviance.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            - power < 0 : Extreme stable distribution. (Requires: preds > 0.)
+            - power = 0 : Normal distribution. (Requires: targets and preds can be any real numbers.)
+            - power = 1 : Poisson distribution. (Requires: targets >= 0 and y_pred > 0.)
+            - 1 < p < 2 : Compound Poisson distribution. (Requires: targets >= 0 and preds > 0.)
+            - power = 2 : Gamma distribution. (Requires: targets > 0 and preds > 0.)
+            - power = 3 : Inverse Gaussian distribution. (Requires: targets > 0 and preds > 0.)
+            - otherwise : Positive stable distribution. (Requires: targets > 0 and preds > 0.)
+
+    Example:
+        >>> from torchmetrics.functional.regression import tweedie_deviance_score
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/tweedie_deviance.py:139]: this is the recommended user import path
```suggestion
        >>> from torchmetrics.functional import tweedie_deviance_score
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+from torchmetrics.metric import Metric
+
+
+class TweedieDevianceScore(Metric):
+    r"""
+    Computes the `Deviance Score <https://en.wikipedia.org/wiki/Tweedie_distribution#The_Tweedie_deviance>`_ between
+    targets and predictions:
+
+    .. math::
+        deviance\_score(\hat{y},y) =
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:32]: have you checked this renders correctly?

File: torchmetrics/regression/tweedie_deviance.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        2 * (\frac{(max(y,0))^{2}}{(1 - power)(2 - power)} - \frac{y(\hat{y})^{1 - power}}{1 - power} + \frac{(\hat{y})
+            ^{2 - power}}{2 - power}), & \text{otherwise}
+        \end{cases}
+
+    where :math:`y` is a tensor of targets values, and :math:`\hat{y}` is a tensor of predictions.
+
+    Forward accepts
+
+    - ``preds`` (float tensor): ``(N,)``
+    - ``targets`` (float tensor): ``(N,)``
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:46]: If I am not wrong, the input can have arbitrary number of extra dimensions 
```suggestion

    - ``preds`` (float tensor): ``(N,...)``
    - ``targets`` (float tensor): ``(N,...)``
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+            default: ``None`` (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When ``None``, DDP
+            will be used to perform the all gather.
+
+    Example:
+        >>> from torchmetrics import TweedieDevianceScore
+        >>> targets = torch.tensor([1.0, 2.0, 3.0, 4.0])
+        >>> preds = torch.tensor([4.0, 3.0, 2.0, 1.0])
+        >>> deviance_score = TweedieDevianceScore(power=0)
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:73]: please use another power than 0 in the examples, as for `power=0` this metric is exactly equal to the mean squared error
```suggestion
        >>> deviance_score = TweedieDevianceScore(power=1)
```

File: torchmetrics/regression/tweedie_deviance.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+        self.add_state("sum_deviance_score", torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("num_observations", torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds: Tensor, targets: Tensor) -> None:  # type: ignore
+        """Update metric states with predictions and targets.
+
+        Args:
+            preds: Predicted tensor with shape ``(N,d)``
+            targets: Ground truth tensor with shape ``(N,d)``
```
Comment (SkafteNicki) [inline torchmetrics/regression/tweedie_deviance.py:108]: inconsistency is shape docstring (probably also in the functional)
```suggestion
            preds: Predicted tensor with shape ``(N,...)``
            targets: Ground truth tensor with shape ``(N,...)``
```

File: tests/regression/test_tweedie_deviance.py [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+    def test_deviance_scores_functional(self, preds, targets, power, sk_metric):
+        self.run_functional_metric_test(
+            preds,
+            targets,
+            tweedie_deviance_score,
+            partial(sk_metric, power=power),
+            metric_args=dict(power=power),
+        )
+
```
Comment (SkafteNicki) [inline tests/regression/test_tweedie_deviance.py:78]: missing tests
```suggestion
    def test_deviance_scores_half_support_cpu(...)
    def test_deviance_scores_half_support_gpu(...)
    def test_deviance_scores_differentiability(...)
```
similar to https://github.com/PyTorchLightning/metrics/blob/e38cb704ecc64b9113fc7703cb90bbea52852afc/tests/regression/test_pearson.py#L72-L84

File: tests/regression/test_tweedie_deviance.py [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    targets=torch.rand(NUM_BATCHES, BATCH_SIZE),
+)
+
+_single_target_inputs2 = Input(
+    preds=torch.rand(NUM_BATCHES, BATCH_SIZE),
+    targets=torch.rand(NUM_BATCHES, BATCH_SIZE),
+)
+
+
+def _sk_deviance(preds: Tensor, targets: Tensor, power: int):
```
Comment (SkafteNicki) [inline tests/regression/test_tweedie_deviance.py:42]:
```suggestion
def _sk_deviance(preds: Tensor, targets: Tensor, power: float):
```

File: tests/regression/test_tweedie_deviance.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+
+_single_target_inputs1 = Input(
+    preds=torch.rand(NUM_BATCHES, BATCH_SIZE),
+    targets=torch.rand(NUM_BATCHES, BATCH_SIZE),
+)
+
+_single_target_inputs2 = Input(
+    preds=torch.rand(NUM_BATCHES, BATCH_SIZE),
+    targets=torch.rand(NUM_BATCHES, BATCH_SIZE),
+)
```
Comment (SkafteNicki) [inline tests/regression/test_tweedie_deviance.py:39]: if the metric supports additional dimensions, we should also test that
```suggestion
)

_multi_target_inputs = Input(
    preds=torch.rand(NUM_BATCHES, BATCH_SIZE, 5),
    targets=torch.rand(NUM_BATCHES, BATCH_SIZE, 5),
)
```

File: torchmetrics/regression/tweedie_deviance.py (hunk 15 above)
Comment (karthikrangasai) [reply torchmetrics/regression/tweedie_deviance.py:32]: Yes I had checked, it does render correctly.

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Review by Borda — APPROVED** [review]
